### ISSUE TITLE
Feature Request: CNAME / Domain name for Buckets

### ISSUE THREAD

**Issue by steffen-mueller** [anomalyco/sst#5123]
### Feature Request

**Context**: Currently, deploying a frontend works fine with the StaticSite etc. components. But this always comes with a CDN distribution. If you want to use the bucket directly, you have to either live with a generic S3 URL or roll your own domain and SSL handling.

**Proposal**: Add the ability to add a `customBucketDomain` for static sites and its friend components. It behaves similarly to the `customDomain` property, but just applies to the bucket, not the cloudfront distribution. The same mechanic should work on the normal bucket component.

### Why should anybody want to do that?

 We have encountered 3 cases that force us to sidestep the CDN in our production app:

**CDN vs geo-blocking firewalls**: Surprisingly, enterprise customers like to use firewalls that block IPs by countries / geo-location. Even "normal" european countries like Italy are sometimes blocked - the block lists are distributed by cybersecurity providers. These enterprise customers could not load our frontend on some machines, depending on their location, because CDN edge nodes were blocked. And because it hits (part of) countries like Italy as well, the coarse price-level settings for CloudFront (you can limit the CDN to US+Europe) doesn't cut it. Even without this rather weird geo-blocking thing, we have some enterprise customers that whitelist based on IPs. 

**Consequent Data Residency Feature**
Our app supports data residency, meaning that customers can choose one of dozens of data centers where their data is stored and processed. Each availability zone runs an own instance of the app. For the sake of clear communication, we want the frontend to be loaded from the same data center where the app is running. Otherwise you have discussions like "Yeah, it runs exclusively in Metropolis. Your data never leaves that. Well, except for some parts of the app that are loaded from one of hundreds of global edge nodes. But don't worry,  they are static. But yes, they interact with your data. Yes exactly, we might load code from Bangalore that interacts with your data stored in Metropolis. We don't control that. We are still compliant, aren't we?"

**Upload Buckets**
We let users upload media files to a S3 bucket. We'd love to add a custom domain to that bucket directly, so the file domain visible to users is "uploads.app.com" and not the gibberish S3 bucket URL.

Thanks for considering this!

**Comment by jayair** [comment]
We have some ideas around this. We want to let the Router to be able to replace the CDN that the site uses. But this needs a bit of work on our side.

**Comment by steffen-mueller** [comment]
Hi @jayair and @fwang, is there any update or schedule for the changes you plan for the "router can replace CDN" thing?

**Comment by jayair** [comment]
Yeah so now you can use a separate CDN for SSR and static sites: https://sst.dev/blog/frontends-are-hard#a-new-approach

**Comment by steffen-mueller** [comment]
Woha, that's amazing. Missed that blog post. Thanks, we'll use that ASAP ;)

**Comment by steffen-mueller** [comment]
Dang, while writing up our ticket for the implementation I realized:

While this helps in putting together nice URL schemes with functions and buckets etc., **the result still stays as CloudFront distribution**. This means that out of my 3 reasons for the change request, two still stay open:

* **CDN vs geo-blocking firewalls**: the `priceClass` distribution property stays the only way to limit the entry points. Our enterprise customers still can't work with that, the router does not help here.
* Same with the "**consequent data residency**": the served bucket may be stored on the US east coast, but be processed and cached in Italy by the CloudFront distribution.

**Summary**: Please reconsider my original proposal as the router does not help here at all.
